### ISSUE TITLE
Run a Docker registry

### ISSUE THREAD

**Issue by thockin** [kubernetes/kubernetes#1319]
@jbeda's idea.

It might be useful to have our standard set up scripts have the master run a registry. This would make it easy to upload/distribute "local" images across the cluster.

While Docker has a clear idea around local and remote images, with k8s we now have "cluster local" images that need to be distributed fast across the cluster but don't need to bounce off of the public registry.

**Comment by pires** [comment]
Interesting discussion going on here. Here's my two cents.

Having a registry per master when there's only one master seems to suffice, but what happens when kubernetes gets support for multiple masters? Even on a active/standby scenario, this could be troublesome, don't you agree?

Now, optionally running a kube service where you can easily publish/cache your private and public Docker images, that is available to every node on the cluster, sounds a terrific idea - even if it's a SPOF.

**Comment by smarterclayton** [comment]
It doesn't have to be spof - we've been running multiple copies of dumb registries behind an auth layer under a service, talking to an object store.  Unless you were referring to the Kube master as being SPOF today, which it is (for mutation to the cluster), which is why post 1.0 ensuring its available through duplication will be a priority.

**Comment by pires** [comment]
@smarterclayton who's we? And is there any documentation on that set-up that you can share?

**Comment by thockin** [comment]
I think this is the right model, with a selection of different registry
implementations (local storage, GCS, etc).  There's a little issue that we
need to sort sort thru regarding DNS from nodes (outside containers).

**Comment by bgrant0607** [comment]
cc @dchen1107 @ArtfulCoder

**Comment by ddysher** [comment]
@thockin @bgrant0607 

Anyone takes this in gsoc?  Trying to find another feature in k8s to work on.

**Comment by bgrant0607** [comment]
@ddysher No, this issue is available.

**Comment by ddysher** [comment]
ok, self-assign

**Comment by preillyme** [comment]
+1 I think this is a great idea. If no one from gsoc takes it we can work on it at Kismatic.

**Comment by preillyme** [comment]
I'll work on this with @ddysher so that we don't duplicate the work.

**Comment by mimizone** [comment]
@preillyme what would be your approach? put the registry components in containers on the master, sort of as current registries are being deployed? 
We have a recent interest in running a registry in full distributed mode in kubernetes as pods. It seems like it could fit nicely in the k8 model. What do you think?

**Comment by bgrant0607** [comment]
Yes, I recommend making it an add-on, similar to DNS:
https://github.com/GoogleCloudPlatform/kubernetes/tree/master/cluster/addons/dns

**Comment by ddysher** [comment]
Sorry, I've been distracted quite a bit last month.

Making it an add-on is ideal and preferrable, but I think we need support from various components to make it work.  Suppose we created pods running a stock docker/registry image, and a service `kube-registry` for this registry addon.  To push/pull a user image `registry.default.kubernetes.local/user/image:tag`, we need:

1. A way to push the image from local repo to cluster registry
Option1: Make kubectl smart enough to push the image, e.g. `kubectl create image registry.default.kubernetes.local/user/image:tag`.  But how will this work? I think a common pattern today is to have master proxy the request, but proxying the image sounds a lot of work.
Option2: Have a different backing storage, like GCS, and we just push image to that storage.  This depends on some external services, I'd like to avoid it first.  e.g. user can use local disk by default for small workload.
2. A way to pull the image:
We need to give the registry service a fixed IP, similar to dns.  When pulling image, kubelet itself translates registry.default.kubernetes.local to the fixedIP

There are more to consider, authentification, caching, backend storage, which we could provide a similar config like docker registry does.  Particularly for backend storage, it would be nice if we can integrate docker registry options with PV.

**Comment by smarterclayton** [comment]
> On May 9, 2015, at 2:03 PM, Deyuan Deng <notifications@github.com> wrote:
> 
> Sorry, I've been distracted quite a bit last month.
> 
> Making it an add-on is ideal and preferrable, but I think we need support from various components to make it work. Suppose we created pods running a stock docker/registry image, and a service kube-registry for this registry addon. To push/pull a user image registry.default.kubernetes.local/user/image:tag, we need:
> 
> A way to push the image from local repo to cluster registry Option1: Make kubectl smart enough to push the image, e.g. kubectl create image registry.default.kubernetes.local/user/image:tag. But how will this work? I think a common pattern today is to have master proxy the request, but proxying the image sounds a lot of work. Option2: Have a different backing storage, like GCS, and we just push image to that storage. This depends on some external services, I'd like to avoid it first. e.g. user can use local disk by default for small workload.

We use the service ip in Openshift - the hosts have access to it.  We don't wire the nodes to use the cluster DNS yet but it's a small step to that and it will give you the IP you wantz

For disk, that's the default we use in the Openshift registry setup - runs the image on local disk first, then you switch to object store or gluster or NFS or gce disk when ready.

> A way to pull the image: We need to give the registry service a fixed IP, similar to dns. When pulling image, kubelet itself translates registry.default.kubernetes.local to the fixedIP
We'll probably move DNS to the Kube-proxy, at which point it's not an issue.  A fixed IP makes sense mostly so you have have the docker daemon trust that ip.  But eventually we'll just want the ability to sign the CA for the cluster and distribute it to the nodes in a way that docker accepts.
> There are more to consider, authentification, caching, backend storage, which we could provide a similar config like docker registry does. Particularly for backend storage, it would be nice if we can integrate docker registry options with PV.
> 
Yeah, that's on our list in terms of being able to easily connect a PV to an RC or deployment object so you can do the transition from simple disk to PV

**Comment by ddysher** [comment]
On Sat, May 9, 2015 at 5:14 PM Clayton Coleman <notifications@github.com>
wrote:

>
>
> > On May 9, 2015, at 2:03 PM, Deyuan Deng <notifications@github.com>
> wrote:
> >
> > Sorry, I've been distracted quite a bit last month.
> >
> > Making it an add-on is ideal and preferrable, but I think we need
> support from various components to make it work. Suppose we created pods
> running a stock docker/registry image, and a service kube-registry for this
> registry addon. To push/pull a user image
> registry.default.kubernetes.local/user/image:tag, we need:
> >
> > A way to push the image from local repo to cluster registry Option1:
> Make kubectl smart enough to push the image, e.g. kubectl create image
> registry.default.kubernetes.local/user/image:tag. But how will this work? I
> think a common pattern today is to have master proxy the request, but
> proxying the image sounds a lot of work. Option2: Have a different backing
> storage, like GCS, and we just push image to that storage. This depends on
> some external services, I'd like to avoid it first. e.g. user can use local
> disk by default for small workload.
>
> We use the service ip in Openshift - the hosts have access to it. We don't
> wire the nodes to use the cluster DNS yet but it's a small step to that and
> it will give you the IP you wantz
>

Being able to access the service IP in all hosts is essential to distribute
images across k8s; kubelet can just pull from that IP for images.  l'm
trying to figure out the best way to upload an image.  There is no cluster
'build service', so client needs to push an image to the registry.  I'm not
sure if new service model could help here, in which case we might be able
to make it external accessible.  Our last resort would be to proxy through
master, which should work on all cluster setup.

@smarterclayton @thockin Any idea?  Is there anything I'm not aware of that
solves the problem nicely, or maybe it's just easier to offer a 'build
service' that takes a dockerfile/image spec, build it and store it in
registry?


> For disk, that's the default we use in the Openshift registry setup - runs
> the image on local disk first, then you switch to object store or gluster
> or NFS or gce disk when ready.
>

I'm interested to know how can we switch storage on-the-fly.  At minimal,
we'll need to change the configuration and restart the registry?


> > A way to pull the image: We need to give the registry service a fixed
> IP, similar to dns. When pulling image, kubelet itself translates
> registry.default.kubernetes.local to the fixedIP
> We'll probably move DNS to the Kube-proxy, at which point it's not an
> issue. A fixed IP makes sense mostly so you have have the docker daemon
> trust that ip. But eventually we'll just want the ability to sign the CA
> for the cluster and distribute it to the nodes in a way that docker
> accepts.
>

SG.  Insecure registry is probably a good start for now.

**Comment by smarterclayton** [comment]
> On May 9, 2015, at 8:25 PM, Deyuan Deng <notifications@github.com> wrote:
> 
> On Sat, May 9, 2015 at 5:14 PM Clayton Coleman <notifications@github.com>
> wrote:
> 
> >
> >
> > > On May 9, 2015, at 2:03 PM, Deyuan Deng <notifications@github.com>
> > wrote:
> > >
> > > Sorry, I've been distracted quite a bit last month.
> > >
> > > Making it an add-on is ideal and preferrable, but I think we need
> > support from various components to make it work. Suppose we created pods
> > running a stock docker/registry image, and a service kube-registry for this
> > registry addon. To push/pull a user image
> > registry.default.kubernetes.local/user/image:tag, we need:
> > >
> > > A way to push the image from local repo to cluster registry Option1:
> > Make kubectl smart enough to push the image, e.g. kubectl create image
> > registry.default.kubernetes.local/user/image:tag. But how will this work? I
> > think a common pattern today is to have master proxy the request, but
> > proxying the image sounds a lot of work. Option2: Have a different backing
> > storage, like GCS, and we just push image to that storage. This depends on
> > some external services, I'd like to avoid it first. e.g. user can use local
> > disk by default for small workload.
> >
> > We use the service ip in Openshift - the hosts have access to it. We don't
> > wire the nodes to use the cluster DNS yet but it's a small step to that and
> > it will give you the IP you wantz
> >
> 
> Being able to access the service IP in all hosts is essential to distribute
> images across k8s; kubelet can just pull from that IP for images. l'm
> trying to figure out the best way to upload an image. There is no cluster
> 'build service', so client needs to push an image to the registry. I'm not
> sure if new service model could help here, in which case we might be able
> to make it external accessible. Our last resort would be to proxy through
> master, which should work on all cluster setup.

You can make an outside registry internally available if you want, but I don't know that you *have* to have the registry in the cluster. If you have the registry CA on every node, the docker daemon can query the internal DNS you're set.  It should be easier to get CAs into nodes but that's something that can be solved outside of Kube for now.

> 
> @smarterclayton @thockin Any idea? Is there anything I'm not aware of that
> solves the problem nicely, or maybe it's just easier to offer a 'build
> service' that takes a dockerfile/image spec, build it and store it in
> registry?

Well... Openshift has one of those that's designed to be integrated into a Kube deployment... So yes :). However, someone pushing images into the Kube cluster makes a lot of sense, but something has to be able to translate "new image available" into "do something with that image" (that's really the Deployments concept we are trying to move from Openshift into Kube).

> 
> 
> > For disk, that's the default we use in the Openshift registry setup - runs
> > the image on local disk first, then you switch to object store or gluster
> > or NFS or gce disk when ready.
> >
> 
> I'm interested to know how can we switch storage on-the-fly. At minimal,
> we'll need to change the configuration and restart the registry?

And copy the data, which could be something your registry does (if two volumes exist, the first volume is empty but the second has data, copy the second set of data into the first).  However, the registry will work best if scaled horizontally, so object store is a great complement to the registry (having an easy to use object store example on the system would be nice).

**Comment by ddysher** [comment]
> You can make an outside registry internally available if you want, but I don't know that you *have* to have the registry in the cluster. If you have the registry CA on every node, the docker daemon can query the internal DNS you're set.  It should be easier to get CAs into nodes but that's something that can be solved outside of Kube for now.

Isn't the pain point now is that people have to have an external registry to deploy their application?  How much difference does it make between using dockerhub vs registry+CAs?   BTW, I haven't really dug into securities in k8s, so probably miss some background.

> Well... Openshift has one of those that's designed to be integrated into a Kube deployment... So yes :). However, someone pushing images into the Kube cluster makes a lot of sense, but something has to be able to translate "new image available" into "do something with that image" (that's really the Deployments concept we are trying to move from Openshift into Kube).

Yeah, thinking more, I also think pusing images makes more sense.  That's the deploy process that docker tries to enforce (or maybe other possible image spec);  user will otherwise have to push source code instead.

Now what's the 'something'?  I think what we really want is the ability to bring up a built-in service that's externally accessible. dns addon doesn't have the problem as it's used internally.  I've mentioned master as proxy, but it's just one way to do that.  We can also create a 'imagecontroller', or maybe use service publicIPs.  What do you think?

**Comment by pmorie** [comment]
@ncdc

**Comment by ddysher** [comment]
@ncdc Maybe your experience from working on streaming library can help here?

**Comment by smarterclayton** [comment]
He's out on PTO and will be back on Monday.

**Comment by ddysher** [comment]
Ah, ok, thanks for the heads up.

**Comment by ncdc** [comment]
@ddysher would you mind summarizing what you're looking to do?

**Comment by ddysher** [comment]
@ncdc Thanks for jumping in.  To answer your question, I'm looking to setup a cluster private registry in k8s :)

Basically, I was thinking that for uploading image from client to the registry, we use the same mechanism you did for kubectl exec, i.e. master as a reverse proxy to the actual location of the pod (here registry service).  I'm a little concerned about the performance though - not sure about the load on master.

FYI, this is what I think a client can do:

1. Add a new subcommand in kubectl to push image:  `kubectl image push {image}`.
  1. Image format should be [REGISTRYHOST/][NAMESPACE/]NAME[:TAG].
    1. If no REGISTRYHOST is given, use default registry, e.g. `registry.default.kubernetes.local`
    2. if no NAMESPACE is given, use namespace in current context
    3. if no TAG is given, use `latest`
  2. Client needs to be aware of the format and tag the image properly.
2. Registry storage is just local disk at first; we can later provide facilities to modify backend storage or bring up a new registry.

**Comment by lavalamp** [comment]
I don't think we need to add a kubectl command; just expose the registry (e.g. via external load balancer) and use docker regularly. We already have pull secrets actually, so that should Just Work.

**Comment by ddysher** [comment]
Ah, pull secrets from this one #7894?  I'll take a look.

**Comment by bgrant0607** [comment]
The registry should be run as an "addon", similar to DNS.

**Comment by smarterclayton** [comment]
The nodes have to have DNS as a resolver, or the registry has to have serving certs matching the service ip.  Or the registry can be run insecure, but that makes it a bit harder.

**Comment by lavalamp** [comment]
I think the order has to be:

1. Make the registry service object.
2. Make a cert matching DNS name and/or external IP of registry service.
3. Stuff cert into a secret.
4. Launch RC that starts a docker registry configured to pull its cert from the secret.

It would be super nice if 2&3 were something that was done for you automatically-- using apiserver's cert to sign...

**Comment by smarterclayton** [comment]
> On May 28, 2015, at 2:36 PM, Daniel Smith <notifications@github.com> wrote:
> 
> I think the order has to be:
> 
> Make the registry service object.
> Make a cert matching DNS name and/or external IP of registry service.
> Stuff cert into a secret.
> Launch RC that starts a docker registry configured to pull its cert from the secret.
> It would be super nice if 2&3 were something that was done for you automatically-- using apiserver's cert to sign...
> 
Yeah, we sketched that out but would be post 1.0 at this point.

**Comment by ddysher** [comment]
I'm slowly learning docker registry and k8s secrets, external service; for now, I think we still have some problems to solve for bootstrapping and usage (not to mention storage, etc).

1. From the official [registry guide](https://github.com/docker/distribution/blob/master/docs/deploying.md#configure-tls-on-a-registry-server), it seems credentials are supposed to be stuffed into the image at build time, and then you just run the image where you build it (or push to a private registry first).  I'm not sure if we have such use case before, I think we can run an unsecure registry behind nginx.  But I do hope there is a way to put credential to the registry at runtime.
2. I like to run the registry as an addon from day one, but it might be hard to use if we make no changes to kubectl.  We need to put k8s ca into docker daemon's certs folder, so things becomes manual when you change a host.  Since we already have ca in our kubeconfig, I think having a lightweight `kubectl push/pull` will be very handy to use.  We can use [docker's client library](https://github.com/docker/docker/tree/master/api/client)
3. Can you make a cert with both domain and IP, use subjectaltname?
4. We'd need to take care of environment where external loadbalancer is not available.  NodePort is probably a good place to start.

**Comment by ddysher** [comment]
EDIT: scratch the docker client library pointer, it's client of docker daemon.   It does seem to be quite a bit of work to incorporate push/pull.

**Comment by uluyol** [comment]
Following the discussion from earlier today, for running the registry itself, I see three modes of operation:

1. Running a registry on every node backed by cloud storage.
1. Running a registry on a central node backed by cloud storage with proxies on every node.
1. Running a registry on a central node backed by a PD with proxies on every node.

The proxies in cases 2 & 3 are needed to sanely use TLS (since DNS is not available from the docker daemon).

For case 1, we'd definitely want to create our own registry image (or fix docker's) since it's a 550 MB image. With case 2 this would be helpful but not as necessary. For both cases 1 and 2, we'd be depending on the storage backends supported by docker (currently regular filesystems, S3, Azure, and rados; the gcs driver found [here](https://registry.hub.docker.com/u/google/docker-registry/) appears to be for docker v1 registry).

Cases 2 & 3 have a single point of failure, whereas case 1 is immune to this. Case 1 should scale the best, case 2 would have additional network overhead, and case 3 would have further overhead (assuming the docker registry has efficient backend implementations).

However, case 3 is the simplest setup in terms of storage, and cases 1 & 2 are equally more difficult.

**Comment by uluyol** [comment]
Now that #12833 is merged, we should talk about the next steps:

1. We should build our own image of the docker registry. The standard image is huge and all we should really need is a static Go binary.
2. The image for the docker registry should be pushed to the nodes. Right now we download it which partly defeats the purpose of having one.
3. The docker registry should request a cert from the apiserver and use that to communicate using TLS. The proxies should be configured to use it. This is blocked by #11725.
4. The docker registry should be backed by cloud storage rather than a PD. There is no GCS support in the v2 docker registry (there is a backend for the v1 registry).
5. Once the docker registry is backed by a PD, we can have an instance of it running on every node and eliminate the proxies. This gives us better scalability and availability. The question of what to do on clusters without cloud storage is unanswered. This eliminates the need for TLS (3).
6. We should push images into the cluster registry and pull from there so we eliminate our dependency on gcr.io.

**Comment by drocsid** [comment]
@uluyol I'd like to test running the docker registry on a bare metal kubernetes environment without cloud storage. I searched for a release with your commit in place, but could not find one.Is there a build available with your registry commit in place, so I could test?

**Comment by uluyol** [comment]
The registry needs to be backed by some sort of persistent storage, so you would need to use NFS, iSCSI, or something similar. The config currently only supports GCE, though this isn't hard to change. I don't know if this has made its way into a release. I also don't know if anyone is actively working on this; I am preoccupied with other things (namely school).

**Comment by steffenmllr** [comment]
@drocsid I'm also looking for a way to run the docker registry on a bare metal / vagrant setup - did you fin a solution for this ?

**Comment by thockin** [comment]
If you change the specification to not use a persistentVolumeClaim and
instead use emptyDir{} - is that sufficient?

**Comment by drocsid** [comment]
@steffenmllr not yet. Nobody above responded regarding a build that contains this registry for testing. I'm in #google-containers , and will be trying to figure this out for the next couple days.

**Comment by drocsid** [comment]
I have found that the commit is part of the alpha release, but not 1.0.5. I'm going to try deploying an alpha cluster and will try to bring the registry up.

**Comment by drocsid** [comment]
I have brought up the registry as suggested by @thockin , using emptyDir{} and removing the PVC. Then I try to push to said registry:

```
sudo docker push 172.16.54.7:5000/hello-world
Error response from daemon: invalid registry endpoint https://172.16.54.7:5000/v0/: unable to ping registry endpoint https://172.16.54.7:5000/v0/
v2 ping attempt failed with error: Get https://172.16.54.7:5000/v2/: tls: oversized record received with length 20527
 v1 ping attempt failed with error: Get https://172.16.54.7:5000/v1/_ping: tls: oversized record received with length 20527. If this private registry supports only HTTP or HTTPS with an unknown CA certificate, please add `--insecure-registry 172.16.54.7:5000` to the daemon's arguments. In the case of HTTPS, if you have access to the registry's CA certificate, no need for the flag; simply place the CA certificate at /etc/docker/certs.d/172.16.54.7:5000/ca.crt
```
I see a reference to a certificate the images folder. Then I also see https://github.com/kubernetes/kubernetes/issues/11725 where it is mentioned that is possible to copy the generated certificate in place. This sounds like the easiest approach until #11725 is closed. @uluyol is this what you did to test your registry? Can you be explicit about which locations I need to place the certificate, the image file and also in the docker certs.d shown above? Anywhere else?

**Comment by uluyol** [comment]
On every node there is a proxy running (gcr.io/google_containers/kube-registry-proxy, the config for this is [kube-registry-proxy.yaml](https://github.com/kubernetes/kubernetes/blob/master/cluster/saltbase/salt/kube-registry-proxy/kube-registry-proxy.yaml)). If you use localhost:5000 on one of the nodes the registry should work fine. You can't push from your machine at home.

If you want that functionality you'll need to take some further steps: create a proxy that serves over SSL instead of plain HTTP (that you can access from outside the cluster), set up a certificate that you trust on your machine or get one signed by a CA, and use the proxy to push images. I don't recommend doing this.

**Comment by drocsid** [comment]
I haven't been trying to push from a machine at home. I will look further at the proxy.yaml. Thanks

**Comment by drocsid** [comment]
Hi @uluyol, the included rc does not seem to use the gcr.io/google_containers/kube-registry-proxy image, but instead a registry image. Then the ha_proxy components are not deployed.

**Comment by drocsid** [comment]
After deploying the registry / proxy / service, I see that I can connect to the registry from localhost. It also appears I can reach the registry through the proxy endpoint from the containers,  telnet://kube-registry.kube-system.svc.cluster.local:5000/ . My current desire is to be able to insert an image into the registry from the host machines running the containers. Then I'd like to be able to deploy pods from that image across the cluster. 

From the host itself outside of the cluster I only seem to be able to connect to localhost:5000, I cannot seem to reach the proxy. Then I'm not sure if I can just push an image to the registry on localhost:5000, and then deploy a pod using that same image across the cluster. I will try to test this out further tomorrow.

**Comment by pwais** [comment]
@drocsid are you by chance pushing from the master node?  I've had related problems doing that in the past.  

Also, have you tried pushing from inside a pod's container?  I.e. can you push to `kube-registry.kube-system.svc.cluster.local:5000` (or even just `kube-registry:5000`) from within a container?  Note that you might need to mount the docker socket inside the container to use the pod's docker daemon, e.g.:
```
...
securityContext:
      privileged: true
...
volumeMounts:
      - name: dockersock
        mountPath: /var/run/docker.sock
```

FWIW in my experience the registry is very sensitive to the hostname used in the `push` command (even when connectivity can be verified OK via telnet).

Super excited to see the cluster registry, especially if it can be easily adapted to not depend on cloud storage.  Thank you @uluyol for hacking on this!

**Comment by uluyol** [comment]
@drocsid you should be running both the proxy rc and the per-node proxies, though it sounds like you've managed this. Now that we have the (experimental) DaemonSet, someone may be able to look into using that instead of managing the pods through salt (this still leaves much of the salt code around unfortunately).

Assuming you have the proxies running, you should be able to do what you've described. The pod config I posted should expose port 5000 on the host making it available to the host docker. I'm not sure what you mean by "host outside of the cluster".

**Comment by paralin** [comment]
Has there been any work on this?

I wanted to try running a registry by using a PersistentVolume attached to a single node replication controller, and then connecting to it via a service. I wanted to just use an internal service via kube-proxy rather than use a load balancer or nodeport. But kube-dns doesn't seem to be active on the actual machine layer where Docker does the pulls, so this doesn't seem possible. 

NodePort might work with binding just to localhost or firewalling off the port, and then connecting to the registry at localhost:port. Then with a PersistentVolume you can have the registry sort of roam over the cluster as needed. Since the actual registry image is hosted on some third party registry - Docker Hub - the cluster should be able to actually fetch the registry image even if the registry doesn't exist already.

**Comment by drocsid** [comment]
I finally got a chance to work on this again. @thockin I followed your documentation for deploying the registry, thanks. I made an effort to get the docker binary running inside a container, and built an image inside of it.

Then came the push attempt.

```bash
jenkins@jenkins-ijphv:/tmp/search_api$ docker tag search-api localhost:5000/search-api
jenkins@jenkins-ijphv:/tmp/search_api$ docker push localhost:5000/search-api
The push refers to a repository [localhost:5000/search-api] (len: 1)
9a503c2f95a5: Image successfully pushed 
bfe7135523a3: Image successfully pushed 
f32d74699997: Image successfully pushed 
4d71802439ee: Image successfully pushed 
6b08f3db1a5d: Pushing [==================================================>] 242.1 MB/242.1 MB
Received unexpected HTTP status: 504 Gateway Time-out
```

**Comment by drocsid** [comment]
It seems that the localhost:5000 endpoint might be visible from the docker host, but not from inside the containers. Then I used the service DNS inside the container, but still seem to have issues connecting. @pwais do you have any suggestions? 

jenkins@jenkins-ijphv:/$ docker tag 9a503c2f95a5 kube-registry:5000/search-api            
jenkins@jenkins-ijphv:/$ docker push kube-registry:5000/search-api
The push refers to a repository [kube-registry:5000/search-api] (len: 1)
unable to ping registry endpoint https://kube-registry:5000/v0/
v2 ping attempt failed with error: Get https://kube-registry:5000/v2/: dial tcp: lookup kube-registry: Temporary failure in name resolution
 v1 ping attempt failed with error: Get https://kube-registry:5000/v1/_ping: dial tcp: lookup kube-registry: Temporary failure in name resolution
jenkins@jenkins-ijphv:/$ curl https://kube-registry:5000/v0/
curl: (35) error:140770FC:SSL routines:SSL23_GET_SERVER_HELLO:unknown protocol
jenkins@jenkins-ijphv:/$ curl -v https://kube-registry:5000/v0/
* Hostname was NOT found in DNS cache
*   Trying 192.168.4.98...
* Connected to kube-registry (192.168.4.98) port 5000 (#0)
* successfully set certificate verify locations:
*   CAfile: none
  CApath: /etc/ssl/certs
* SSLv3, TLS handshake, Client hello (1):
* error:140770FC:SSL routines:SSL23_GET_SERVER_HELLO:unknown protocol
* Closing connection 0
curl: (35) error:140770FC:SSL routines:SSL23_GET_SERVER_HELLO:unknown protocol

**Comment by yujuhong** [comment]
@drocsid, I don't think the [kube-registry add-on](https://github.com/thockin/kubernetes/blob/0b50b7e00f6b9fa29b10ede75adc2460eaad708b/cluster/addons/registry/README.md) works for your use case. The addons explicitly says 
```
The private registry runs as a Pod in your cluster. It does not currently support SSL or authentication, which triggers Docker's "insecure registry" logic. To work around this, we run a proxy on each node in the cluster, exposing a port onto the node (via a hostPort), which Docker accepts as "secure", since it is accessed by localhost.
```

The proxy pod makes it possible for docker to talk to `localhost:5000` **no the node**, instead of `kube-registry.kube-system.svc.cluster.local` to bypass SSL authentication. If you want to use this private registry from within your container, the easiest workaround I can think of is to run the kube-registry-proxy container alongside with your own container in the same pod. Just remove the `hostPort` from the [spec](https://github.com/thockin/kubernetes/blob/0b50b7e00f6b9fa29b10ede75adc2460eaad708b/cluster/addons/registry/README.md#expose-the-registry-on-each-node) since it's no longer needed.

**Comment by drocsid** [comment]
Hi @yujuhong ,

Thanks for your suggestion. I considered co-locating my container with the kube-registry pod itself, since I believe it can talk to localhost:5000. I will either try this or try co-locating with the proxy.

Also, in the registry documentation readme ( under addons/registry ) it states:

```If you are building an image on your Kubernetes Node, you can spell out localhost:5000 when you build and push.``` This is confusing and contradicts what you are saying above. Finally, later down the road it might be worthwhile to try to get the SSL certs working or similar, then maybe we could push to kube-registry, instead of localhost:5000.

Thanks again.

**Comment by yujuhong** [comment]
> If you are building an image on your Kubernetes Node, you can spell out localhost:5000 when you build and push. This is confusing and contradicts what you are saying above "no(t) the node". Finally, later down the road it might be worthwhile to try to get the SSL certs working or similar, then maybe we could push to kube-registry, instead of localhost:5000. If I can get a registry working, I might have time to try extending it.

Sorry I corrected my typo. I meant you could access `localhost:5000` ***on*** any node in your cluster where the `kube-registry-proxy` is running. But since you are running docker within a pod/container, this doesn't work for you. Co-locating your pod with a proxy container may be more convenient because you can create multiple pods to push images and delete them whenever you want without affecting the pod hosting the registry. It is really up to your use case though.

Yes, enabling SSL is one of the future improvements listed at the end of the documents. It'd be great if you can extend this :-)

**Comment by drocsid** [comment]
So far I have tried using co-locating my container with the kube-registry pod itself. Then from my container, I can connect to the registry, and verify from the registry container logs.

```jenkins@kube-registry-v1-17nb9:/$  curl -v telnet://localhost:5000     
* Rebuilt URL to: telnet://localhost:5000/
* Hostname was NOT found in DNS cache
*   Trying ::1...
* Connected to localhost (::1) port 5000 (#0)
^C
jenkins@kube-registry-v1-17nb9:/$  curl -v localhost:5000         
* Rebuilt URL to: localhost:5000/
* Hostname was NOT found in DNS cache
*   Trying ::1...
* Connected to localhost (::1) port 5000 (#0)
> GET / HTTP/1.1
> User-Agent: curl/7.38.0
> Host: localhost:5000
> Accept: */*
> 
< HTTP/1.1 404 Not Found
< Content-Type: text/plain; charset=utf-8
< Docker-Distribution-Api-Version: registry/2.0
< Date: Thu, 19 Nov 2015 01:28:58 GMT
< Content-Length: 19
< 
404 page not found
* Connection #0 to host localhost left intact
jenkins@kube-registry-v1-17nb9:/$ 
```

```
::1 - - [19/Nov/2015:01:28:58 +0000] "GET / HTTP/1.1" 404 19 "" "curl/7.38.0"
```

However when I try to push I get the following error, and don't see the connection attempt being made in the registry logs:

```
jenkins@kube-registry-v1-17nb9:/$ docker push localhost:5000/search-api
The push refers to a repository [localhost:5000/search-api] (len: 1)
Sending image list
Error: Status 503 trying to push repository search-api: "<html><body><h1>503 Service Unavailable</h1>\nNo server is available to handle this request.\n</body></html>\n"
```

Then I suspect there is another issue at hand.

**Comment by drocsid** [comment]
I have now mounted the docker binary and socket inside the registry container itself, to see if I can push from there. I have stopped the proxy, just to test from inside that container. Then either there is some issue with using the docker binary and socket from inside the container, or another issue with the registry itself.

```
root@kube-registry-v2-3bgnt:/# curl -v localhost:5000
* Rebuilt URL to: localhost:5000/
* Hostname was NOT found in DNS cache
*   Trying ::1...
* Connected to localhost (::1) port 5000 (#0)
> GET / HTTP/1.1
> User-Agent: curl/7.35.0
> Host: localhost:5000
> Accept: */*
> 
< HTTP/1.1 404 Not Found
< Content-Type: text/plain; charset=utf-8
< Docker-Distribution-Api-Version: registry/2.0
< Date: Thu, 19 Nov 2015 19:44:41 GMT
< Content-Length: 19
< 
404 page not found
* Connection #0 to host localhost left intact
root@kube-registry-v2-3bgnt:/# curl -v 127.0.0.1:5000
* Rebuilt URL to: 127.0.0.1:5000/
* Hostname was NOT found in DNS cache
*   Trying 127.0.0.1...
* Connected to 127.0.0.1 (127.0.0.1) port 5000 (#0)
> GET / HTTP/1.1
> User-Agent: curl/7.35.0
> Host: 127.0.0.1:5000
> Accept: */*
> 
< HTTP/1.1 404 Not Found
< Content-Type: text/plain; charset=utf-8
< Docker-Distribution-Api-Version: registry/2.0
< Date: Thu, 19 Nov 2015 19:44:53 GMT
< Content-Length: 19
< 
404 page not found
* Connection #0 to host 127.0.0.1 left intact
```

```
root@kube-registry-v2-3bgnt:/# docker push localhost:5000/drocsid/nginx
The push refers to a repository [localhost:5000/drocsid/nginx] (len: 1)
Sending image list
Put http://localhost:5000/v1/repositories/drocsid/nginx/: dial tcp 127.0.0.1:5000: getsockopt: connection refused
```

**Comment by drocsid** [comment]
I believe I have isolated my registry issues to be with using the docker binary inside of the container. I had tried to adapt the suggestions:  http://jpetazzo.github.io/2015/09/03/do-not-use-docker-in-docker-for-ci/ for my registry container. However I cannot push, but I can perform other operations inside the container. For example just now:
```root@kube-registry-v2-3bgnt:/# docker stop 74f710a47ad9
74f710a47ad9```

**Comment by drocsid** [comment]
I have found my issues pushing to the registry were with linked libraries in the docker containers. After volume mounting the system libraries, and using export LD_LIBRARY_PATH="/var/lib64/", I am able to push to the registry.

**Comment by bgrant0607** [comment]
Some discussion here:
https://github.com/kubernetes/kubernetes/pull/16010#issuecomment-158150897
@gabrtv

**Comment by drocsid** [comment]
I may have jumped the gun. While I can push to the registry from within my pod, not all of the cluster hosts can pull from it. So far, I can only pull from it on the host that has the registry pod running. Then there is some issue with the proxy or docker itself..

```shell
core@coreos003
docker pull localhost:5000/search_api:test1                                                                                                                                                                          
Pulling repository localhost:5000/search_api
Error: Status 503 trying to pull repository search_api: "<html><body><h1>503 Service Unavailable</h1>\nNo server is available to handle this request.\n</body></html>\n"
```

```shell
core@coreos004 ~ $ docker pull localhost:5000/search_api:test1                                                                                                                                                                          
test1: Pulling from search_api

1565e86129b8: Already exists 
a604b236bcde: Already exists 
5822f840e16b: Already exists 
00ef84872481: Already exists 
d3bc5121beff: Already exists 
d65af916d98d: Already exists 
6bcee452d13c: Already exists 
a450fff6ccf1: Already exists 
4266f385037c: Already exists 
aa416f0dcbce: Already exists 
76bc28326eb8: Already exists 
7559fba81cb9: Already exists 
db73c08bca61: Already exists 
77f8e2db858f: Already exists 
cde27d083ab4: Already exists 
77ce940f2164: Already exists 
8f8b40e8db11: Already exists 
845499caa9a4: Already exists 
03ae6afabf34: Already exists 
1eedacba997c: Already exists 
de6696b7661d: Already exists 
503e1670ebf4: Already exists 
0b06a8f0db82: Already exists 
1cc0a3c68aee: Already exists 
0381f674cd4f: Already exists 
87cdd9d445da: Already exists 
27f2f0b5c606: Already exists 
003ff525f7e9: Already exists 
91ff4a2e0c7a: Already exists 
bdcafc3ffa14: Already exists 
Digest: sha256:aeea2ab79995e7a4f0fadd8f9c430fef66d84870e92355595119c96e6b6204ab
Status: Image is up to date for localhost:5000/search_api:test1
```

**Comment by drocsid** [comment]
After restarting kubelet, and the proxy, I did not seem to be running into this issue anymore. Then the registry seems to work end to end, across my cluster.

**Comment by cognitiaclaeves** [comment]
@drocsid , can you share the setup that you are using? I'm looking for a way to deploy containers to a private registry that can be pulled by all the nodes in a cluster.

I was thinking that, ultimately, I'd like to deploy the registry through the cluster itself as a kubernetes service and have the nodes in the cluster be configured with the service IP ( the registry would be configured to use some kind of persistence mechanism located on another server )

... however ... by the looks of the length of this thread, this may be a lot a very difficult ( or maybe even impossible? ) thing to do.

So, I'd like to know exactly what you did get working end-to-end, so I can figure out if your setup will solve our most pressing issue: deploying custom docker containers through kubernetes without using public registries.

**Comment by thockin** [comment]
@freehan

**Comment by smarterclayton** [comment]
It's not that difficult, but does require some  configuration on the node.
First, secure or insecure?  If secure, you'll need server certs signed with
the cluster DNS name of your service, and each node has to have that ca in
its trust chain so docker can see it (you can sign the service cluster ip
but then you would take downtime on a cert change.  Then the nodes have to
have the cluster DNS resolver set.

Otherwise, docker --insecure-registry can be used against the service ip
cidr.

In Openshift the on cluster registry runs using persistent volumes for many
people (over NFS, be sure to set source affinity on your service because of
NFS visibility propagation), or more scalably, against object store.  We
use normal deployments to scale out the registry.



On Dec 4, 2015, at 7:33 PM, Jason <notifications@github.com> wrote:

@drocsid <https://github.com/drocsid> , can you share the setup that you
are using? I'm looking for a way to deploy containers to a private registry
that can be pulled by all the nodes in a cluster.

I was thinking that, ultimately, I'd like to deploy the registry through
the cluster itself as a kubernetes service and have the nodes in the
cluster be configured with the service IP ( the registry would be
configured to use some kind of persistence mechanism located on another
server )

... however ... by the looks of the length of this thread, this may be a
lot a very difficult ( or maybe even impossible? ) thing to do.

So, I'd like to know exactly what you did get working end-to-end, so I can
figure out if your setup will solve our most pressing issue: deploying
custom docker containers through kubernetes without using public registries.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/issues/1319#issuecomment-162117885>
.

**Comment by drocsid** [comment]
@cognitiaclaeves As suggested, first make sure the DNS addon is deployed and working on your cluster. Then you would deploy the registry addon after. Finally, what are your goals for using the registry. I wanted to build containers from within containers, and therefore had some more difficulties. Feel free to contact me via email, and I'm often on the #kubernetes-users slack channel if you want to discuss more.

**Comment by cognitiaclaeves** [comment]
It turned out that it was easier to just deploy a v2 registry container
<https://docs.docker.com/registry/deploying/>. ( I had previously been
trying to get a v1 registry working out of a VM. )

The reason for the private registry is to give kubernetes nodes a means for
running images in a protected subnet. The way it is now, I pull images I
need from outside registries, tag and push them into the private registry,
and then configure the yaml files to pull those tagged images from the
private registry when needed.

( I think I'm saying all of that correctly ... )

**Comment by bprashanth** [comment]
This is probably not making 1.2, we should continue discussion of what's in scope for 1.3 on https://github.com/kubernetes/kubernetes/issues/19033

**Comment by bgrant0607** [comment]
A registry is available as an addon. Closing as obsolete.

**Comment by metrue** [comment]
I'd like to share my the way (https://github.com/metrue/fx/issues/315) I'm doing in [fx](https://github.com/metrue/fx)
